checker/src/main/java/org/checkerframework/checker/determinism/DeterminismAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,851 @@
+            // (Example: Ordering.natural() in tests/all-systems/PolyCollectorTypeVars.java)
+            // For static methods, receiverType is the AnnotatedTypeMirror of the class in which the
+            // invoked method "node" is declared.
+            if (receiverType == null) {
+                return super.visitMethodInvocation(node, annotatedRetType);
+            }
+
+            ExecutableElement m = TreeUtils.elementFromUse(node);
+
+            // If return type (non-array, non-collection, and non-iterator) resolves to
+            // @OrderNonDet, replaces the annotation on the return type with @NonDet.
+            if (annotatedRetType.hasAnnotation(ORDERNONDET)
+                    && !mayBeOrderNonDet(annotatedRetType)) {
+                annotatedRetType.replaceAnnotation(NONDET);
+            }
+
+            // Annotates the return type of "equals()" method called on a Set receiver
+            // as described in the specification of this method.
+
+            // Example1: @OrderNonDet Set<@OrderNonDet List<@Det Integer>> s1;
+            //           @OrderNonDet Set<@OrderNonDet List<@Det Integer>> s2;
+            // s1.equals(s2) is @NonDet
+
+            // Example 2: @OrderNonDet Set<@Det List<@Det Integer>> s1;
+            //            @OrderNonDet Set<@Det List<@Det Integer>> s2;
+            // s1.equals(s2) is @Det
+            // TODO-rashmi: this can be more precise (@Det receiver and @OrderNonDet argument)
+            TypeElement receiverUnderlyingType =
+                    TypesUtils.getTypeElement(receiverType.getUnderlyingType());
+
+            // Without this check, NullPointerException in Collections class with buildJdk.
+            // Likely cause: Collections has a private constructor?
+            // Error at line: public class Collections {
+            // TODO-rashmi: check why?
+            if (receiverUnderlyingType == null) {
+                return super.visitMethodInvocation(node, annotatedRetType);
+            }
+
+            if (isEqualsMethod(m)) {
+                AnnotatedTypeMirror argument = getAnnotatedType(node.getArguments().get(0));
+                if (isSubClassOf(receiverType, setInterfaceTypeMirror)
+                        && receiverType.hasAnnotation(ORDERNONDET)
+                        && !hasOrderNonDetListAsTypeArgument(receiverType)
+                        && isSubClassOf(argument, setInterfaceTypeMirror)
+                        && argument.hasAnnotation(ORDERNONDET)
+                        && !hasOrderNonDetListAsTypeArgument(argument)) {
+                    annotatedRetType.replaceAnnotation(DET);
+                }
+            }
+
+            // The following code is a workaround for Issue#14
+            // (https://github.com/t-rasmud/checker-framework/issues/14).
+            // Checks if the return type is not a TYPEVAR, and if the invoked method belongs
+            // to a set of (hardcoded) Collection methods in the JDK that return a generic type.
```

Comment:
"A generic type is a generic class or interface that is parameterized over types." [See Java tutorial](https://docs.oracle.com/javase/tutorial/java/generics/types.html).

Do you mean type variable?